network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,348 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! The handshake module implements the handshake part of the protocol.
+//! This module also implements additional anti-DoS mitigation,
+//! by including a timestamp in each handshake initialization message.
+//! Refer to the module's documentation for more information.
+//! A successful handshake returns a `NoiseStream` which is defined in [socket] module.
+//!
+//! [socket]: network::noise_wrapper::socket
+
+use futures::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use std::{
+    collections::HashMap,
+    io,
+    sync::{Arc, RwLock},
+    time,
+};
+
+use libra_config::config::NetworkPeerInfo;
+use libra_crypto::{noise, x25519};
+use libra_types::PeerId;
+use netcore::transport::ConnectionOrigin;
+
+use crate::noise_wrapper::session::NoiseStream;
+
+/// In a mutually authenticated network, a client message is accompanied with a timestamp.
+/// This is in order to prevent replay attacks, where the attacker does not know the client's static key,
+/// but can still replay a handshake message in order to force a peer into performing a few Diffie-Hellman key exchange operations.
+///
+/// Thus, to prevent replay attacks a responder will always check if the timestamp is strictly increasing,
+/// effectively considering it as a stateful counter.
+///
+/// If the client timestamp has been seen before, or is not strictly increasing,
+/// we can abort the handshake early and avoid heavy Diffie-Hellman computations.
+/// If the client timestamp is valid, we store it.
+#[derive(Default)]
+pub struct AntiReplayTimestamps(HashMap<x25519::PublicKey, u64>);
+
+impl AntiReplayTimestamps {
+    /// Returns true if the timestamp has already been observed for this peer
+    /// or if it's an old timestamp
+    pub fn is_replay(&self, pubkey: x25519::PublicKey, timestamp: u64) -> bool {
+        if let Some(last_timestamp) = self.0.get(&pubkey) {
+            &timestamp <= last_timestamp
```

Comment:
I'm new to rust.  Why not dereference last_timestamp and compare u64s vs &u64s?

Reply:
I'm actually not sure why I need to borrow here, I was expecting rust to do it automatically, but I wouldn't expect borrowing or dereferencing to change how anything gets compiled eventually.